for i in `cat list`; do sh STAR_2_pass.sh /share/data5/SingleCellSeq/GSE75688/fastq/$i'_1.fastq.gz' /share/data5/SingleCellSeq/GSE7568i 30 /home/liukuai/GSE75688/genomeDir/ /share/data0/reference/Genome/hg19/hg19.chr.fa /home/liukuai/single_cell/homo_annotation_hg19/gencode.v19.annotation.gtf; done

for i in `cat list`; do echo $i >file.txt; 

for i in `cat list`; do echo $i'.Aligned.sortedByCoord.out.bam' >file.txt; samtools mpileup -B -d10000 -f /share/data0/reference/Genome/hg19/hg19.chr.fa -q 40 -b file.txt|python ../../hamimzafar-monovar/src/monovar.py -p 0.002 -a 0.2 -t 0.05 -m 40 -f /share/data0/reference/Genome/hg19/hg19.chr.fa -b file.txt -o ../vcf/$i.vcf;done

samtools mpileup -B -d10000 -f /share/data0/reference/Genome/hg19/hg19.chr.fa -q 40 -b file.txt|python hamimzafar-monovar/src/monovar.py -p 0.002 -a 0.2 -t 0.05 -m 40 -f /share/data0/reference/Genome/hg19/hg19.chr.fa -b file.txt -o output1.vcf

seqkit fq2fa G1.fastq -o g1.fasta



split_libraries_fastq.py -i per_sample_clean/G1/G1.fastq,per_sample_clean/G10/G10.fastq,per_sample_clean/G11/G11.fastq,per_sample_clean/G12/G12.fastq,per_sample_clean/G13/G13.fastq,per_sample_clean/G14/G14.fastq,per_sample_clean/G15/G15.fastq,per_sample_clean/G16/G16.fastq,per_sample_clean/G17/G17.fastq,per_sample_clean/G18/G18.fastq,per_sample_clean/G19/G19.fastq,per_sample_clean/G2/G2.fastq,per_sample_clean/G20/G20.fastq,per_sample_clean/G21/G21.fastq,per_sample_clean/G22/G22.fastq,per_sample_clean/G23/G23.fastq,per_sample_clean/G24/G24.fastq,per_sample_clean/G25/G25.fastq,per_sample_clean/G26/G26.fastq,per_sample_clean/G27/G27.fastq,per_sample_clean/G28/G28.fastq,per_sample_clean/G29/G29.fastq,per_sample_clean/G3/G3.fastq,per_sample_clean/G30/G30.fastq,per_sample_clean/G31/G31.fastq,per_sample_clean/G4/G4.fastq,per_sample_clean/G5/G5.fastq,per_sample_clean/G6/G6.fastq,per_sample_clean/G7/G7.fastq,per_sample_clean/G8/G8.fastq,per_sample_clean/G9/G9.fastq,per_sample_clean/Y1/Y1.fastq,per_sample_clean/Y10/Y10.fastq,per_sample_clean/Y11/Y11.fastq,per_sample_clean/Y12/Y12.fastq,per_sample_clean/Y13/Y13.fastq,per_sample_clean/Y14/Y14.fastq,per_sample_clean/Y15/Y15.fastq,per_sample_clean/Y16/Y16.fastq,per_sample_clean/Y17/Y17.fastq,per_sample_clean/Y18/Y18.fastq,per_sample_clean/Y19/Y19.fastq,per_sample_clean/Y2/Y2.fastq,per_sample_clean/Y20/Y20.fastq,per_sample_clean/Y21/Y21.fastq,per_sample_clean/Y22/Y22.fastq,per_sample_clean/Y23/Y23.fastq,per_sample_clean/Y24/Y24.fastq,per_sample_clean/Y25/Y25.fastq,per_sample_clean/Y26/Y26.fastq,per_sample_clean/Y27/Y27.fastq,per_sample_clean/Y28/Y28.fastq,per_sample_clean/Y29/Y29.fastq,per_sample_clean/Y3/Y3.fastq,per_sample_clean/Y30/Y30.fastq,per_sample_clean/Y31/Y31.fastq,per_sample_clean/Y4/Y4.fastq,per_sample_clean/Y5/Y5.fastq,per_sample_clean/Y6/Y6.fastq,per_sample_clean/Y7/Y7.fastq,per_sample_clean/Y8/Y8.fastq,per_sample_clean/Y9/Y9.fastq -o sl/ --store_demultiplexed_fastq -q 19 -n 100 --barcode_type not-barcoded --sample_ids per_sample_clean/G1/G1.fastq,per_sample_clean/G10/G10.fastq,per_sample_clean/G11/G11.fastq,per_sample_clean/G12/G12.fastq,per_sample_clean/G13/G13.fastq,per_sample_clean/G14/G14.fastq,per_sample_clean/G15/G15.fastq,per_sample_clean/G16/G16.fastq,per_sample_clean/G17/G17.fastq,per_sample_clean/G18/G18.fastq,per_sample_clean/G19/G19.fastq,per_sample_clean/G2/G2.fastq,per_sample_clean/G20/G20.fastq,per_sample_clean/G21/G21.fastq,per_sample_clean/G22/G22.fastq,per_sample_clean/G23/G23.fastq,per_sample_clean/G24/G24.fastq,per_sample_clean/G25/G25.fastq,per_sample_clean/G26/G26.fastq,per_sample_clean/G27/G27.fastq,per_sample_clean/G28/G28.fastq,per_sample_clean/G29/G29.fastq,per_sample_clean/G3/G3.fastq,per_sample_clean/G30/G30.fastq,per_sample_clean/G31/G31.fastq,per_sample_clean/G4/G4.fastq,per_sample_clean/G5/G5.fastq,per_sample_clean/G6/G6.fastq,per_sample_clean/G7/G7.fastq,per_sample_clean/G8/G8.fastq,per_sample_clean/G9/G9.fastq,per_sample_clean/Y1/Y1.fastq,per_sample_clean/Y10/Y10.fastq,per_sample_clean/Y11/Y11.fastq,per_sample_clean/Y12/Y12.fastq,per_sample_clean/Y13/Y13.fastq,per_sample_clean/Y14/Y14.fastq,per_sample_clean/Y15/Y15.fastq,per_sample_clean/Y16/Y16.fastq,per_sample_clean/Y17/Y17.fastq,per_sample_clean/Y18/Y18.fastq,per_sample_clean/Y19/Y19.fastq,per_sample_clean/Y2/Y2.fastq,per_sample_clean/Y20/Y20.fastq,per_sample_clean/Y21/Y21.fastq,per_sample_clean/Y22/Y22.fastq,per_sample_clean/Y23/Y23.fastq,per_sample_clean/Y24/Y24.fastq,per_sample_clean/Y25/Y25.fastq,per_sample_clean/Y26/Y26.fastq,per_sample_clean/Y27/Y27.fastq,per_sample_clean/Y28/Y28.fastq,per_sample_clean/Y29/Y29.fastq,per_sample_clean/Y3/Y3.fastq,per_sample_clean/Y30/Y30.fastq,per_sample_clean/Y31/Y31.fastq,per_sample_clean/Y4/Y4.fastq,per_sample_clean/Y5/Y5.fastq,per_sample_clean/Y6/Y6.fastq,per_sample_clean/Y7/Y7.fastq,per_sample_clean/Y8/Y8.fastq,per_sample_clean/Y9/Y9.fastq

usearch -fastx_uniques reads_uparse.fa -fastaout seqs_unique.fa -minuniquesize 2 -sizeout

 vsearch -derep_fulllength seqs.fna -output seqs_unique
 usearch  -cluster_otus seqs_unique.fa  -otus otus.fa -uparseout uparse.txt -relabel Otu
vsearch  -usearch_global seqs.fna -db result/rep_seqs.fa -otutabout otu_table.txt -biomout otu_table.biom -strand plus -id 0.97 -threads 30